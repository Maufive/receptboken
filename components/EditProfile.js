import React, { Component } from "react";
import axios from "axios";
import Router from "next/router";
import { Loading } from "../components/Loading";
import { StyledForm } from "../components/styles/FormStyles";
import { Button } from "../components/styles/Button";

// Verifiera att användaren som är inloggad är den användare som försöker redigera profilen.
class EditProfile extends Component {
	state = {
		verified: false,
		loading: false,
		photo: this.props.user.photo || null,
		fname: this.props.user.fname,
		lname: this.props.user.lname,
		description: this.props.user.description || ""
	};

	async componentDidMount() {
		await this.verifyUser();
	}

	saveToState = e => {
		this.setState({ [e.target.name]: e.target.value });
	};

	verifyUser = () => {
		// Skickar en postrequest till servern som tar id´t i queryn och id´t som finns på jwt-token och jämför dem på servern
		const user = this.props.user;
		axios
			.post(
				`https://receptboken.herokuapp.com/auth/verify/${this.props.query.id}`,
				{
					user
				}
			)
			.then(response => {
				if (response.data === true) {
					this.setState({ verified: true });
				} else {
					this.props.setMessage(
						"danger",
						"Du har inte tillåtelse att göra detta"
					);
					Router.push("/");
				}
			})
			.catch(error => console.log(error));
	};

	uploadFile = async e => {
		this.setState({ loading: true });
		const files = e.target.files;
		const data = new FormData();
		data.append("file", files[0]);
		data.append("upload_preset", "receptboken_profil");

		const res = await fetch(
			"https://api.cloudinary.com/v1_1/dcgb0fhog/image/upload",
			{
				method: "POST",
				body: data
			}
		).catch(error => console.log(error));
		const file = await res.json();
		this.setState({
			photo: file.secure_url,
			loading: false
		});
	};

	onSubmit = async e => {
		e.preventDefault();
		const data = this.state;
		const id = this.props.user._id;
		await axios
			.post(`${process.env.API}/user/edit/${id}`, { data })
			.then(response => {
				this.props.setMessage("success", response.data);
				Router.push(`/profile?id=${id}`);
				location.reload();
			})
			.catch(error => console.log(error));
	};

	render() {
		return (
			<div>
				<StyledForm onSubmit={this.onSubmit} id="edit-profile-form">
					<h1>Redigera din profil</h1>
					<div>
						<input
							type="text"
							name="fname"
							id="fname"
							onChange={this.saveToState}
							value={this.state.fname}
						/>
						<label htmlFor="fname">Förnamn</label>
					</div>
					<div>
						<input
							type="text"
							name="lname"
							id="lname"
							onChange={this.saveToState}
							value={this.state.lname}
						/>
						<label htmlFor="lname">Efternamn</label>
					</div>
					<div>
						<textarea
							name="description"
							id="description"
							onChange={this.saveToState}
							value={this.state.description}
						/>
						<label
							htmlFor="description"
							style={{
								transform: "translateY(-4.5rem)"
							}}
						>
							Beskrivning
						</label>
					</div>
					<div style={{ flexDirection: "column" }}>
						<p>Ladda upp en profilbild</p>
						<input
							type="file"
							id="file"
							name="photo"
							onChange={this.uploadFile}
							style={{
								margin: "2rem 0"
							}}
						/>
						{this.state.loading && <Loading />}
						{this.state.photo && (
							<img
								width="150"
								id="photo"
								src={this.state.photo}
								alt="Bild förhandsvisning"
								style={{
									borderRadius: "50%"
								}}
							/>
						)}
					</div>
					<Button type="submit" id="submit-button" fullWidth primary>
						Spara ändringar
					</Button>
				</StyledForm>
			</div>
		);
	}
}

export default EditProfile;
