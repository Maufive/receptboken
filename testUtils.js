export const fakeRecept = {
	_id: "1234",
	author: "5678436374re67fwuf76",
	created: "Idag",
	description: ["Steg1", "Steg2", "Steg3"],
	id: "1337",
	ingredients: [
		{
			input: "Lök",
			numberOfUnits: "2",
			units: "dl"
		}
	],
	photo: "asdf123.jpg",
	reviews: [{ score: 1 }],
	servings: "2",
	tags: ["Mat", "Gott"],
	timeRequired: "20 min",
	title: "Mitt jättegoda recept"
};

export const fakeUser = {
	_id: "1234",
	username: "fakeuser@gmail.com",
	password: "1234",
	hearts: [],
	fname: "Niklas",
	lname: "Albinsson",
	description: "Hej jag heter Niklas",
	photo: "http://fakephoto.jpg"
};

export const fakeShoppinglist = [
	{
		_id: "34567876543452176",
		title: "My Shoppinglist",
		created: "12/3 2019",
		author: "61386712376576d7q6ewd65",
		list: [
			{
				input: "Lök",
				numberOfUnits: "5",
				units: "dl"
			},
			{
				input: "Tomater",
				numberOfUnits: "10",
				units: "st"
			}
		]
	}
];
