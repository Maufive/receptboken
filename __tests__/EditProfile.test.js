import React from "react";
import { shallow } from "enzyme";
import toJSON from "enzyme-to-json";
import EditProfile from "../components/EditProfile";
import { fakeUser } from "../testUtils";

const fakeId = {
	id: "1234567890"
};

describe("<EditProfile />", () => {
	let wrapper;
	beforeEach(
		() => (wrapper = shallow(<EditProfile user={fakeUser} query={fakeId} />))
	);

	it("renders without errors and matches snapshot", () => {
		expect(toJSON(wrapper)).toMatchSnapshot();
	});

	//input fname
	describe("Firstname Input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#fname").length).toEqual(1);
		});

		it("should respond to the change event and change the state of the component", () => {
			wrapper.find("#fname").simulate("change", {
				target: { name: "fname", value: "Niklas" }
			});

			expect(wrapper.state("fname")).toEqual("Niklas");
		});
	});
	//input lname
	describe("Lastname Input", () => {
		it("renders the input", () => {
			expect(wrapper.find("#lname").length).toEqual(1);
		});

		it("should respond to the change event and change the state of the component", () => {
			wrapper.find("#lname").simulate("change", {
				target: { name: "lname", value: "Albinsson" }
			});

			expect(wrapper.state("lname")).toEqual("Albinsson");
		});
	});

	//textarea description
	describe("Textarea", () => {
		it("renders the textarea", () => {
			expect(wrapper.find("#description").length).toEqual(1);
		});

		it("should respond to the change event and change the state of the component", () => {
			wrapper.find("#description").simulate("change", {
				target: { name: "description", value: "This is a test description" }
			});
		});
	});

	//input file name="photo"
	describe("File upload", () => {
		it("renders the upload-input", () => {
			expect(wrapper.find("#file").length).toEqual(1);
		});
	});

	//button submit
	describe("Submitbutton", () => {
		it("renders the button", () => {
			expect(wrapper.find("#submit-button").length).toEqual(1);
		});
	});
});
