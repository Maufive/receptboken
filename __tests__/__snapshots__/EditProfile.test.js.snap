// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<EditProfile /> renders without errors and matches snapshot 1`] = `
<div>
  <FormStyles__StyledForm
    id="edit-profile-form"
    onSubmit={[Function]}
  >
    <h1>
      Redigera din profil
    </h1>
    <div>
      <input
        id="fname"
        name="fname"
        onChange={[Function]}
        type="text"
        value="Niklas"
      />
      <label
        htmlFor="fname"
      >
        Förnamn
      </label>
    </div>
    <div>
      <input
        id="lname"
        name="lname"
        onChange={[Function]}
        type="text"
        value="Albinsson"
      />
      <label
        htmlFor="lname"
      >
        Efternamn
      </label>
    </div>
    <div>
      <textarea
        id="description"
        name="description"
        onChange={[Function]}
        value="Hej jag heter Niklas"
      />
      <label
        htmlFor="description"
        style={
          Object {
            "transform": "translateY(-4.5rem)",
          }
        }
      >
        Beskrivning
      </label>
    </div>
    <div
      style={
        Object {
          "flexDirection": "column",
        }
      }
    >
      <p>
        Ladda upp en profilbild
      </p>
      <input
        id="file"
        name="photo"
        onChange={[Function]}
        style={
          Object {
            "margin": "2rem 0",
          }
        }
        type="file"
      />
      <img
        alt="Bild förhandsvisning"
        id="photo"
        src="http://fakephoto.jpg"
        style={
          Object {
            "borderRadius": "50%",
          }
        }
        width="150"
      />
    </div>
    <Button
      fullWidth={true}
      id="submit-button"
      primary={true}
      type="submit"
    >
      Spara ändringar
    </Button>
  </FormStyles__StyledForm>
</div>
`;
